src/Framework/Constraint/Count.php

```diff
@@ -116,4 +81,82 @@ protected function failureDescription($other): string
             $this->expectedCount
         );
     }
+
+    private function getCountOfTraversable(Traversable $traversable): int
+    {
+        if ($this->traversableCounts === null) {
+            $this->traversableCounts = new \SplObjectStorage();
+        }
+
+        while ($traversable instanceof IteratorAggregate) {
+            $traversable = $traversable->getIterator();
+        }
+
+        if ($this->traversableCounts->contains($traversable)) {
+            return $this->traversableCounts[$traversable];
+        }
+
+        if ($traversable instanceof Iterator) {
+            return $this->countIteratorWithoutChangingPositionIfPossible($traversable);
+        }
+
+        return $this->getCountOfTraversableThatIsNotIteratorOrIteratorAggregate($traversable);
+    }
+
+    private function countIteratorWithoutChangingPositionIfPossible(Iterator $iterator): int
+    {
+        $key   = $iterator->key();
+        $count = $this->getCountOfIterator($iterator);
+
+        $this->traversableCounts->attach($iterator, $count);
+
+        if ($this->iteratorIsRewindable($iterator)) {
+            $this->traversableCounts->detach($iterator);
```

Comment:
Maybe it is possible to `attach` only if needed, so that you don't need to `detach`

Reply:
As far as I can tell this is not possible, hence the approach taken.